kedro_neptune/__init__.py

```diff
@@ -267,61 +314,74 @@ def before_pipeline_run(
+                           custom_run_id=self._run_id,
+                           source_files=source_files or None)
+
+        run[INTEGRATION_VERSION_KEY] = __version__
 
-        neptune_run = self._neptune_metadata.load()
+        current_namespace = run[base_namespace]
 
-        log_command(namespace=neptune_run)
-        log_git_sha(namespace=neptune_run)
-        log_run_params(namespace=neptune_run, run_params=run_params)
-        log_data_catalog_metadata(namespace=neptune_run, catalog=catalog)
-        log_pipeline_metadata(namespace=neptune_run, pipeline=pipeline)
+        os.environ.setdefault('NEPTUNE_API_TOKEN', api_token or '')
+        os.environ.setdefault('NEPTUNE_PROJECT', project or '')
+
+        log_command(namespace=current_namespace)
+        log_run_params(namespace=current_namespace, run_params=run_params)
+        log_data_catalog_metadata(namespace=current_namespace, catalog=catalog)
+        log_pipeline_metadata(namespace=current_namespace, pipeline=pipeline)
 
     @hook_impl
     def before_node_run(
             self,
             node: Node,
-            catalog: DataCatalog,
             inputs: Dict[str, Any],
-            is_async: bool,
-            run_id: str,
+            catalog: DataCatalog
     ):
-        neptune_run = self._neptune_metadata.load()
-        current_namespace = neptune_run[f'nodes/{node.short_name}']
+        run = catalog.load('neptune_metadata')
+        current_namespace = run[f'nodes/{node.short_name}']
 
         if inputs:
-            current_namespace['inputs'].log(list(sorted(inputs.keys())))
+            current_namespace['inputs'] = list(sorted(inputs.keys()))
 
         for input_name, input_value in inputs.items():
             if input_name.startswith('params:'):
                 current_namespace[f'parameters/{input_name[len("params:"):]}'] = input_value
 
         os.environ['NEPTUNE_MONITORING_NAMESPACE'] = f'monitoring/nodes/{node.short_name}'
+
         self._node_execution_timers[node.short_name] = time.time()
 
     @hook_impl
     def after_node_run(
             self,
             node: Node,
-            outputs: Dict[str, Any],
-            inputs: Dict[str, Any]
+            catalog: DataCatalog,
+            outputs: Dict[str, Any]
     ) -> None:
+        del os.environ['NEPTUNE_MONITORING_NAMESPACE']
+
         execution_time = float(time.time() - self._node_execution_timers[node.short_name])
 
-        neptune_run = self._neptune_metadata.load()
-        current_namespace = neptune_run[f'nodes/{node.short_name}']
+        run = catalog.load('neptune_metadata')
+        current_namespace = run[f'nodes/{node.short_name}']
 
         if outputs:
-            current_namespace['outputs'].log(list(sorted(outputs.keys())))
+            current_namespace['outputs'] = list(sorted(outputs.keys()))
         current_namespace['execution_time'] = execution_time
 
+        log_data_catalog_metadata(namespace=run, catalog=catalog)
```

Comment:
I really don't doing it this often, but I guess that's architectural discussion (a bit late for that)

Reply:
?

Reply:
if I understand correctly you're calling `log_data_catalog_metadata` (which always uses the same catalog) after each node execution -- and there can be arbitrarily many nodes...

Reply:
DataCatalog is changing over time. For every dataset in data catalog we check two things:
- Does it exists on drive/s3/etc.? -> For example weights of trained model might be saved after some node execution (it didn't exist before pipeline run)
- Does I already uploaded it? -> Same dataset will never be uploaded twice or more into Neptune by checking the existance of namespace.

In general running `log_data_catalog_metadata` at the end of execution of every node fills the requirement that we upload every dataset immedietly after it was created. In case of failure of some node we still have as much of datasets as we can in Neptune.

I do think that main architectural problem here is the creation of neptune Runs every time the node end its execution and checking for it's structure.